include/bluetooth/audio/has.h

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2021 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_SERVICES_HAS_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_SERVICES_HAS_H_
+
+/**
+ * @brief Hearing Access Service (HAS)
+ *
+ * @defgroup bt_has Hearing Access Service (HAS)
+ *
+ * @ingroup bluetooth
+ * @{
+ *
+ * The Hearing Access Service is used to identify a hearing aid and optionally to control hearing
+ * aid presets. This API implements the server functionality.
+ *
+ * [Experimental] Users should note that the APIs can change as a part of ongoing development.
+ */
+
+#include <zephyr/types.h>
+#include <bluetooth/bluetooth.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief The preset record contains all the information about a hearing aid preset. */
+struct bt_has_preset {
+	/** @brief Preset set active callback
+	 *
+	 *  The callback can also be used locally in which case no connection will be set.
+	 *  Once the preset has changed, the bt_has_active_preset_changed stall be called
+	 *  to complete the procedure.
+	 *
+	 *  @param conn   The connection that is requesting to read
+	 *  @param preset The preset that's being set active
+	 *  @param sync   Whether the server must relay this change to the other member of
+	 * 				  the Binaural Hearing Aid Set.
+	 */
+	void (*set_active)(struct bt_conn *conn, const struct bt_has_preset *preset,
+		 bool sync);
+
+	/** @brief Preset name read callback
+	 *
+	 *  The callback can also be used locally in which case no connection will be set.
+	 *
+	 *  @param conn   The connection that is requesting to read
+	 *  @param preset The preset that's name being read
+	 *  @param buf    Buffer to place the read result in
+	 *  @param len    Length of data to read
+	 *
+	 *  @return Number fo bytes read.
+	 */
+	ssize_t (*read_name)(struct bt_conn *conn, const struct bt_has_preset *preset,
+			 void *buf, uint16_t len);
+
+	/** @brief Preset name write callback
+	 * 
+	 *  If not set, the preset name is not writtable.
+	 *
+	 *  @param conn   The connection that is requesting to write
+	 *  @param preset The preset that's name being written
+	 *  @param buf    Buffer with the data to write
+	 *  @param len    Number of bytes in the buffer
+	 *
+	 *  @return Number of bytes written.
+	 */
+	ssize_t	(*write_name)(struct bt_conn *conn, const struct bt_has_preset *preset,
+			 const void *buf, uint16_t len);
```

Comment:
As discussed at the F2F, this sort of API is useful for propagating the control and storage of the data to the upper layers, but will not be great for attempting to qualify the implementation, as the tests will heavily depend on the upper layers (which isn't part of the Zephyr stack, and thus not part of the Zephyr qualification as far as I understand.).

Reply:
> As discussed at the F2F, this sort of API is useful for propagating the control and storage of the data to the upper layers, but will not be great for attempting to qualify the implementation, as the tests will heavily depend on the upper layers (which isn't part of the Zephyr stack, and thus not part of the Zephyr qualification as far as I understand.).

Are you afraid the application will not store the requested value? Regarding qualification, test application can be provided, and that will work.

Reply:
> Are you afraid the application will not store the requested value? Regarding qualification, test application can be provided, and that will work.

Isn't that only if there's an upper tester involved in the test case? 
For something like this, there is no upper tester, and thus if the stack depend on a sample/test application to qualify, and we then remove that code for an actual application, it doesn't really make sense (to me at least) that the qualification is still valid, if you remove the part of the code that does the actual functionality. 

This is the main issue I see with offloading the logic and storage to the application layer. 

Of course, if the HAS API is meant to be used by a HA implementation (that we also want to qualify, this is fine), but then we just can't qualify HAS without HAP (which is also fine I suppose).

Reply:
What in the case if the preset name shall be persistent - so that it's the same when battery is replaced? Should HAP implementation manage the persistent storage for presets? What in the case of FW update so that preset is removed? How we can know when to remove the entry from storage if e.g. dynamic presets are supported?

Reply:
If dynamic preset names are supported, then HAP could manage the persistent storage of the name. That's similar to GAP that currently manages the device name in persistent storage. 

> What in the case of FW update so that preset is removed?  

HAP does not have an requirements towards persistent storage. What HAP should ensure is that if a client write "This is a new preset name" and then reads, the value, then it will read the same value, or be notified if there are other changes to the name inbetween. 

A client can, however, not always assume that the name it wrote will be the same between connections or reboots. 

If we add this logic in a test/sample application, then we can probably ensure that it is correct and even qualify that application. However, if someone wants to create their own application (typical usecase), and they introduce a bug in the callback that e.g. only copies `n - 1` characters (`n` being the number of characters to be written), then the application is non-compliant. In other words, if the stack is not responsible for the logic, then we cannot properly qualify the stack, as we cannot guarantee the behavior of the upper layers.